Mathlib/Tactic/FailIfNoProgress.lean

```diff
@@ -4,63 +4,200 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Thomas Murrills
 -/
 import Lean
+import Std
+import Mathlib.Lean.Meta.Compare
 
 /-!
 # Fail if no progress
 
-This implements the `fail_if_no_progress` tactic, which fails if no actual progress is made by the
-following tactic sequence.
+This implements the `fail_if_no_progress` tactic (and its internal versions), which fail if no
+progress is made by the tactic sequence provided to it.
 
-"Actual progress" means that either the number of goals has changed, that the
-number or presence of expressions in the context has changed, or that the type of some expression
-in the context or the type of the goal is no longer definitionally equal to what it used to be at
-reducible transparency.
-
-This means that, for example, `1 - 1` changing to `0` does not count as actual progress, since
-```lean
-example : (1 - 1 = 0) := by with_reducible rfl
-```
+"Progress" here means a change between the list of goals before the tactics and the list of goals
+afterwards, as specified by the settings in `FailIfNoProgress.Config`. By default, only
+"observable" changes are checked; this file also provides the preset configs `.anyChanges` and
+`.onlyExprs` which check for any changes whatsoever and only compare expressions appearing in the
+target and local context, respectively.
 
 This tactic is useful in situations where we want to stop iterating some tactics if they're not
 having any  effect, e.g. `repeat (fail_if_no_progress simp <;> ring_nf)`.
 
+## Possible future features
+
+* Tracing to show what progress was made when `fail_if_no_progress` succeeds
+  (i.e. `fail_if_no_progress?`)
+
 -/
 
-namespace Mathlib.Tactic
+namespace Mathlib.Tactic.FailIfNoProgress
+
+open Lean Meta Mathlib.Meta Elab Tactic Parser.Tactic
+
+/-- The overall mode for a `FailIfNoProgress` config. If `mode := .normal`, the rest of the config
+is used to determine equality as described. If `mode := .quick`, the rest of the config is
+ignored, and `fail_if_no_progress` only checks if any of the original goals have been assigned. If
+not, no progress is considered to have been made. -/
+inductive Mode where
+/-- Compares the goal lists before and after the tactic using the given config settings. -/
+| normal
+/-- Only checks if the initial metavariables have been assigned or not. If this mode is used, no
+other checks are performed. -/
+| quick
+deriving Repr, Inhabited
+
+/--
+Config data for `failIfNoProgress`, `fail_if_no_progress`, and `runAndFailIfNoProgress`. This
```

Comment:
```suggestion
Config data for `failIfNoProgress`, `fail_if_no_progress`, and `runAndFailIfNoProgress`. This
```
On first reading I found this confusing, because I didn't already at my finger tips the difference between these. Maybe just ordering them as `fail_if_no_progress`, `failIfNoProgress` and `runAndFailIfNoProgress` would be helpful. Or better a parenthesis: "`fail_if_no_progress` (the user facing tactic), `failIfNoProgress` (in `TacticM`)".